:+1: 
transfer commands now work with `-id -to`.

When no transfer is found, however, the error message is a little confusing, by telling first that not all arguments are given:

```
[server1] All params are not set! Need at least (-to -rule -file | -to -id) params

org.waarp.common.database.exception.WaarpDatabaseNoDataException: Transfer not found
	at org.waarp.openr66.database.data.DbTaskRunner.<init>(DbTaskRunner.java:692)
	at org.waarp.openr66.client.AbstractTransfer.getParamsInternal(AbstractTransfer.java:386)
	at org.waarp.openr66.client.AbstractTransfer.getParams(AbstractTransfer.java:301)
	at org.waarp.openr66.client.DirectTransfer.main(DirectTransfer.java:160)
Caused by: org.waarp.openr66.dao.exception.DAONoDataException: No org.waarp.openr66.dao.database.postgres.PostgreSQLTransferDAO found
	at org.waarp.openr66.dao.database.DBTransferDAO.select(DBTransferDAO.java:445)
	at org.waarp.openr66.database.data.DbTaskRunner.<init>(DbTaskRunner.java:687)
	... 3 common frames omitted
```

--- a/WaarpCommon/src/main/java/org/waarp/common/xml/XmlUtil.java
+++ b/WaarpCommon/src/main/java/org/waarp/common/xml/XmlUtil.java
@@ -588,8 +588,8 @@ public final class XmlUtil {
             continue;
           }
           for (final Node element : elts) {
-            final XmlValue[] newValue = read((Element) element,
-                                             decls[i].getSubXml());
+            final XmlValue[] newValue =
+                read((Element) element, decls[i].getSubXml());
             if (newValue == null) {
               continue;
             }
@@ -675,8 +675,8 @@ public final class XmlUtil {
             continue;
           }
           for (final Node element : elts) {
-            final XmlValue[] newValue = read((Element) element,
-                                             decls[i].getSubXml());
+            final XmlValue[] newValue =
+                read((Element) element, decls[i].getSubXml());
             if (newValue == null) {
               continue;
             }

--- a/WaarpR66/src/main/java/org/waarp/openr66/client/AbstractTransfer.java
+++ b/WaarpR66/src/main/java/org/waarp/openr66/client/AbstractTransfer.java
@@ -22,6 +22,7 @@ package org.waarp.openr66.client;
 import org.waarp.common.command.exception.CommandAbstractException;
 import org.waarp.common.database.data.AbstractDbData.UpdatedInfo;
 import org.waarp.common.database.exception.WaarpDatabaseException;
+import org.waarp.common.database.exception.WaarpDatabaseNoDataException;
 import org.waarp.common.logging.SysErrLogger;
 import org.waarp.common.logging.WaarpLogger;
 import org.waarp.common.logging.WaarpLoggerFactory;
@@ -270,6 +271,18 @@ public abstract class AbstractTransfer implements Runnable {
   protected static Timestamp ttimestart;
   protected static boolean snormalInfoAsWarn = true;
 
+  protected static void clear() {
+    rhost = null;
+    localFilename = null;
+    rule = null;
+    fileInfo = null;
+    ismd5 = false;
+    block = 0x10000; // 64K
+    nolog = false;
+    idt = ILLEGALVALUE;
+    ttimestart = null;
+    snormalInfoAsWarn = true;
+  }
   /**
    * Parse the parameter and set current values
    *
@@ -298,6 +311,18 @@ public abstract class AbstractTransfer implements Runnable {
           Messages.getString("Configuration.NeedCorrectConfig")); //$NON-NLS-1$
       return false;
     }
+    return getParamsInternal(args);
+  }
+
+  /**
+   * Internal getParams without configuration initialization, but still as
+   * first argument
+   *
+   * @param args
+   *
+   * @return True if all parameters were found and correct
+   */
+  protected static boolean getParamsInternal(String[] args) {
     // Now set default values from configuration
     block = Configuration.configuration.getBlockSize();
     int i = 1;
@@ -375,9 +400,18 @@ public abstract class AbstractTransfer implements Runnable {
         rule = runner.getRuleId();
         localFilename = runner.getOriginalFilename();
         return true;
+      } catch (final WaarpDatabaseNoDataException e) {
+        logger.error("No transfer found with this id and partner");
+        logger.error(Messages.getString("AbstractBusinessRequest.NeedMoreArgs",
+                                        "(-to -rule -file | -to -id with " +
+                                        "existing id transfer)")
+                     //$NON-NLS-1$
+            , e);
+        return false;
       } catch (final WaarpDatabaseException e) {
         logger.error(Messages.getString("AbstractBusinessRequest.NeedMoreArgs",
-                                        "(-to -rule -file | -to -id)")
+                                        "(-to -rule -file | -to -id) with a " +
+                                        "correct database connexion")
                      //$NON-NLS-1$
             , e);
         return false;

--- a/WaarpR66/src/main/java/org/waarp/openr66/client/DirectTransfer.java
+++ b/WaarpR66/src/main/java/org/waarp/openr66/client/DirectTransfer.java
@@ -61,7 +61,6 @@ public class DirectTransfer extends AbstractTransfer {
   }
 
   /**
-   *
    * @return True if this DirectTransfer should limit the retry of connection
    */
   public boolean isLimitRetryConnection() {
@@ -69,9 +68,8 @@ public class DirectTransfer extends AbstractTransfer {
   }
 
   /**
-   *
    * @param limitRetryConnection True (default) for limited retry on
-   * connection, False to have no limit
+   *     connection, False to have no limit
    */
   public void setLimitRetryConnection(final boolean limitRetryConnection) {
     this.limitRetryConnection = limitRetryConnection;

--- a/WaarpR66/src/main/java/org/waarp/openr66/client/RequestTransfer.java
+++ b/WaarpR66/src/main/java/org/waarp/openr66/client/RequestTransfer.java
@@ -134,8 +134,7 @@ public class RequestTransfer implements Runnable {
         }
         rhost = srequested;
         try {
-          srequester = Configuration.configuration
-              .getHostId(admin.getSession(), srequested);
+          srequester = Configuration.configuration.getHostId(srequested);
         } catch (final WaarpDatabaseException e) {
           logger.error(Messages.getString("RequestTransfer.5") + srequested,
                        e); //$NON-NLS-1$
@@ -149,8 +148,7 @@ public class RequestTransfer implements Runnable {
         }
         rhost = srequester;
         try {
-          srequested = Configuration.configuration
-              .getHostId(admin.getSession(), srequester);
+          srequested = Configuration.configuration.getHostId(srequester);
         } catch (final WaarpDatabaseException e) {
           logger.error(Messages.getString("RequestTransfer.5") + srequester,
                        e); //$NON-NLS-1$

--- a/WaarpR66/src/main/java/org/waarp/openr66/client/SpooledDirectoryTransfer.java
+++ b/WaarpR66/src/main/java/org/waarp/openr66/client/SpooledDirectoryTransfer.java
@@ -488,8 +488,8 @@ public class SpooledDirectoryTransfer implements Runnable {
                 ' ' + filename + ' ';
             try {
               R66Future r66Future = new R66Future(true);
-              final String srequester = Configuration.configuration
-                  .getHostId(admin.getSession(), host);
+              final String srequester =
+                  Configuration.configuration.getHostId(host);
               // Try restart
               final RequestTransfer transaction =
                   new RequestTransfer(r66Future, fileItem.specialId, host,
@@ -502,6 +502,7 @@ public class SpooledDirectoryTransfer implements Runnable {
               r66Future.awaitOrInterruptible();
               // reset fileItem usage
               setValid(fileItem);
+              specialId = fileItem.specialId;
             } catch (final WaarpDatabaseException e) {
               if (admin.getSession() != null) {
                 admin.getSession().checkConnectionNoException();
@@ -534,11 +535,10 @@ public class SpooledDirectoryTransfer implements Runnable {
               if (specialId != ILLEGALVALUE) {
                 // Clean previously transfer if any
                 try {
-                  final String srequester = Configuration.configuration
-                      .getHostId(admin.getSession(), host);
-                  text =
-                      "Request Transfer Cancelled: " + specialId +
-                      ' ' + filename + ' ';
+                  final String srequester =
+                      Configuration.configuration.getHostId(host);
+                  text = "Request Transfer Cancelled: " + specialId + ' ' +
+                         filename + ' ';
                   // Cancel
                   logger.debug("Will try to cancel {}", specialId);
                   final RequestTransfer transaction2 =

--- a/WaarpR66/src/main/java/org/waarp/openr66/database/data/DbTaskRunner.java
+++ b/WaarpR66/src/main/java/org/waarp/openr66/database/data/DbTaskRunner.java
@@ -383,8 +383,7 @@ public class DbTaskRunner extends AbstractDbDataDao<Transfer> {
     originalSize = requestPacket.getOriginalSize();
     setOriginalSizeTransferMap(originalSize);
     // itself but according to SSL
-    pojo.setRequester(
-        Configuration.configuration.getHostId(dbSession, requested));
+    pojo.setRequester(Configuration.configuration.getHostId(requested));
 
     // Retrieve rule
     this.rule = new DbRule(getRuleId());
@@ -684,9 +683,9 @@ public class DbTaskRunner extends AbstractDbDataDao<Transfer> {
     logger.trace("TRACE ID {}", id);
     try {
       transferAccess = DAOFactory.getInstance().getTransferDAO();
-      pojo = transferAccess
-          .select(id, Configuration.configuration.getHostId(), requested,
-                  Configuration.configuration.getHostId());
+      String requester = Configuration.configuration.getHostId(requested);
+      pojo = transferAccess.select(id, requester, requested,
+                                   Configuration.configuration.getHostId());
     } catch (final DAOConnectionException e) {
       throw new WaarpDatabaseException(e);
     } catch (final DAONoDataException e) {
@@ -792,7 +791,8 @@ public class DbTaskRunner extends AbstractDbDataDao<Transfer> {
         Configuration.configuration.getR66Mib().notifyInfoTask(
             "Task is " + pojo.getUpdatedInfo().name(), this);
       } else {
-        logger.debug("Could send a SNMP trap here since {}", pojo.getUpdatedInfo());
+        logger.debug("Could send a SNMP trap here since {}",
+                     pojo.getUpdatedInfo());
       }
     } else {
       if (pojo.getGlobalStep() != Transfer.TASKSTEP.TRANSFERTASK ||

--- a/WaarpR66/src/main/java/org/waarp/openr66/protocol/configuration/Configuration.java
+++ b/WaarpR66/src/main/java/org/waarp/openr66/protocol/configuration/Configuration.java
@@ -1408,21 +1408,34 @@ public class Configuration {
   }
 
   /**
-   * @param dbSession
    * @param remoteHost
    *
    * @return the HostId according to remoteHost (and its SSL status)
    *
    * @throws WaarpDatabaseException
    */
-  public String getHostId(DbSession dbSession, String remoteHost)
-      throws WaarpDatabaseException {
+  public String getHostId(String remoteHost) throws WaarpDatabaseException {
     final DbHostAuth dbHostAuth = new DbHostAuth(remoteHost);
     try {
       return configuration.getHostId(dbHostAuth.isSsl());
     } catch (final OpenR66ProtocolNoSslException e) {
       throw new WaarpDatabaseException(e);
     }
+  }
+
+  /**
+   * @param dbSession
+   * @param remoteHost
+   *
+   * @return the HostId according to remoteHost (and its SSL status)
+   *
+   * @throws WaarpDatabaseException
+   * @deprecated Use getHostId(String remoteHost)
+   */
+  @Deprecated
+  public String getHostId(DbSession dbSession, String remoteHost)
+      throws WaarpDatabaseException {
+    return getHostId(remoteHost);
   }
 
   private static class UsageStatistic extends TimerTask {

--- a/WaarpR66/src/main/java/org/waarp/openr66/protocol/localhandler/LocalChannelReference.java
+++ b/WaarpR66/src/main/java/org/waarp/openr66/protocol/localhandler/LocalChannelReference.java
@@ -156,10 +156,6 @@ public class LocalChannelReference {
    * PartnerConfiguration
    */
   private volatile PartnerConfiguration partner;
-  /**
-   * DbSession for Database that do not support concurrency in access
-   */
-  private volatile DbSession noconcurrencyDbSession;
 
   /**
    * @param networkChannelRef
@@ -280,9 +276,6 @@ public class LocalChannelReference {
    * @return the actual dbSession
    */
   public DbSession getDbSession() {
-    if (noconcurrencyDbSession != null) {
-      return noconcurrencyDbSession;
-    }
     if (networkServerHandler != null) {
       return networkServerHandler.getDbSession();
     }

--- a/WaarpR66/src/main/java/org/waarp/openr66/protocol/localhandler/ServerActions.java
+++ b/WaarpR66/src/main/java/org/waarp/openr66/protocol/localhandler/ServerActions.java
@@ -2618,16 +2618,13 @@ public class ServerActions extends ConnectionActions {
       throw new OpenR66ProtocolNotAuthenticatedException(
           "Not authenticated while Information received");
     }
-    final DbSession dbSession =
-        localChannelReference != null? localChannelReference.getDbSession() :
-            admin.getSession();
     String remote = session.getAuth().getUser();
     if (jsonOutput && remoteHost != null && !remoteHost.isEmpty()) {
       remote = remoteHost;
     }
     String local;
     try {
-      local = Configuration.configuration.getHostId(dbSession, remote);
+      local = Configuration.configuration.getHostId(remote);
     } catch (final WaarpDatabaseException e1) {
       logger.error("Remote Host is unknown", e1);
       throw new OpenR66ProtocolNoDataException("Remote Host is unknown", e1);

--- a/WaarpR66/src/test/java/org/waarp/openr66/protocol/junit/NetworkClientTest.java
+++ b/WaarpR66/src/test/java/org/waarp/openr66/protocol/junit/NetworkClientTest.java
@@ -37,6 +37,7 @@ import org.junit.runners.MethodSorters;
 import org.openqa.selenium.By;
 import org.openqa.selenium.NoSuchElementException;
 import org.openqa.selenium.StaleElementReferenceException;
+import org.waarp.common.database.DbConstant;
 import org.waarp.common.database.DbPreparedStatement;
 import org.waarp.common.database.exception.WaarpDatabaseException;
 import org.waarp.common.database.exception.WaarpDatabaseNoConnectionException;
@@ -45,8 +46,6 @@ import org.waarp.common.database.exception.WaarpDatabaseSqlException;
 import org.waarp.common.digest.FilesystemBasedDigest;
 import org.waarp.common.file.FileUtils;
 import org.waarp.common.logging.SysErrLogger;
-import org.waarp.common.logging.WaarpLogLevel;
-import org.waarp.common.logging.WaarpLoggerFactory;
 import org.waarp.common.utility.Processes;
 import org.waarp.common.utility.Version;
 import org.waarp.openr66.client.Message;
@@ -66,6 +65,7 @@ import org.waarp.openr66.database.data.DbHostAuth;
 import org.waarp.openr66.database.data.DbRule;
 import org.waarp.openr66.database.data.DbTaskRunner;
 import org.waarp.openr66.protocol.configuration.Configuration;
+import org.waarp.openr66.protocol.configuration.Messages;
 import org.waarp.openr66.protocol.exception.OpenR66ProtocolBusinessException;
 import org.waarp.openr66.protocol.http.rest.handler.DbConfigurationR66RestMethodHandler;
 import org.waarp.openr66.protocol.http.rest.handler.DbHostAuthR66RestMethodHandler;
@@ -543,9 +543,8 @@ public class NetworkClientTest extends TestAbstract {
     assertEquals(0, spooledThread.spooledDirectoryTransfer.getError());
   }
 
-  private void test_Spooled(final SpooledThread spooledThread,
-                            final int factor) throws IOException,
-                                                    InterruptedException {
+  private void test_Spooled(final SpooledThread spooledThread, final int factor)
+      throws IOException, InterruptedException {
     final int size = 200;
     Configuration.configuration.changeNetworkLimit(0, 0, 0, 0, 1000);
     File directory = new File(SpooledThread.TMP_R_66_TEST_OUT_EXAMPLE);
@@ -569,9 +568,9 @@ public class NetworkClientTest extends TestAbstract {
     Thread.sleep(4000 / factor);
     logger.warn("Third file");
     generateOutFile(
-        SpooledThread.TMP_R_66_TEST_OUT_EXAMPLE + "/testTaskBig.txt", size+1,
+        SpooledThread.TMP_R_66_TEST_OUT_EXAMPLE + "/testTaskBig.txt", size + 1,
         "abcdefghij");
-    Thread.sleep(3000 / factor);
+    Thread.sleep(4000 / factor);
     logger.warn("Third file deleted");
     totestBig.delete();
     Thread.sleep(2000 / factor);
@@ -621,8 +620,8 @@ public class NetworkClientTest extends TestAbstract {
   }
 
   @Test
-  public void test7_SpooledWrongHost() throws IOException,
-                                              InterruptedException {
+  public void test7_SpooledWrongHost()
+      throws IOException, InterruptedException {
     logger.warn("Start Test of Spooled Retry Transfer");
     //WaarpLoggerFactory.setLogLevel(WaarpLogLevel.DEBUG);
     SpooledThread spooledThread = new SpooledThread();
@@ -649,7 +648,8 @@ public class NetworkClientTest extends TestAbstract {
   }
 
   @Test
-  public void test7_SpooledWrongHostIgnore() throws IOException, InterruptedException {
+  public void test7_SpooledWrongHostIgnore()
+      throws IOException, InterruptedException {
     logger.warn("Start Test of Spooled Ignored Changed File Transfer");
     SpooledThread spooledThread = new SpooledThread();
     spooledThread.ignoreAlreadyUsed = true;
@@ -737,6 +737,120 @@ public class NetworkClientTest extends TestAbstract {
     executorService.shutdown();
     totest.delete();
     assertEquals("Success should be total", nb, success);
+    assertEquals("Errors should be 0", 0, error);
+  }
+
+  static class SubmitTransferTest extends SubmitTransfer {
+
+    public SubmitTransferTest(final R66Future future, final String remoteHost,
+                              final String filename, final String rulename,
+                              final String fileinfo, final boolean isMD5,
+                              final int blocksize, final long id,
+                              final Timestamp starttime) {
+      super(future, remoteHost, filename, rulename, fileinfo, isMD5, blocksize,
+            id, starttime);
+    }
+
+    static public boolean checkInternalParams(String[] args) {
+      return getParamsInternal(args);
+    }
+
+    static public void clearInternal() {
+      clear();
+    }
+    static public String getParamRule() {
+      return rule;
+    }
+
+    static public String getParamLocalFilename() {
+      return localFilename;
+    }
+  }
+
+  @Test
+  public void test5_DirectTransferThroughId() throws Exception {
+    final File totest = generateOutFile("/tmp/R66/out/testTask.txt", 10);
+    logger.warn("Start Test of DirectTransfer");
+    final long time1 = System.currentTimeMillis();
+    int success = 0;
+    int error = 0;
+    long specialId = DbConstant.ILLEGALVALUE;
+    // First transfer
+    {
+      final R66Future future = new R66Future(true);
+      final TestTransferNoDb transaction =
+          new TestTransferNoDb(future, "hostas", "testTask.txt", "rule3",
+                               "Test SendDirect Small", true, 8192,
+                               DbConstantR66.ILLEGALVALUE, networkTransaction);
+      transaction.run();
+      future.awaitOrInterruptible();
+      if (future.getRunner() != null) {
+        dbTaskRunners.add(future.getRunner());
+        specialId = future.getRunner().getSpecialId();
+      }
+      if (future.isSuccess()) {
+        success++;
+        logger.warn("Success for first transfer");
+      } else {
+        error++;
+        fail("Should be able to find out the previous DbTaskRunner");
+      }
+    }
+    // Now second transfer based on ID
+    String[] args =
+        { "falsConfigFile", "-id", "" + specialId, "-to", "hostas" };
+    if (SubmitTransferTest.checkInternalParams(args)) {
+      String rule = SubmitTransferTest.getParamRule();
+      String localFilename = SubmitTransferTest.getParamLocalFilename();
+      logger.warn("Success for finding previous transfer {} {} {}", specialId,
+                  rule, localFilename);
+      final R66Future future2 = new R66Future(true);
+      final Timestamp newstart = new Timestamp(System.currentTimeMillis());
+      logger.warn("Start second submit transfer");
+      SubmitTransfer submitTransfer =
+          new SubmitTransfer(future2, "hostas", localFilename, rule,
+                             "Test Send 2 Submit Small", true, 8192,
+                             DbConstantR66.ILLEGALVALUE, newstart);
+      logger.warn("Running second submit transfer");
+      submitTransfer.run();
+      logger.warn("Waiting second submit transfer");
+      future2.awaitOrInterruptible();
+      logger
+          .warn("End wait for second submit transfer {}", future2.isSuccess());
+      if (future2.isSuccess()) {
+        success++;
+      } else {
+        error++;
+        fail("Should be able to find out the previous DbTaskRunner");
+      }
+      logger.warn("Prepare transfer Success: " + success + " Error: " + error);
+      assertEquals("Success should be 2", 2, success);
+      assertEquals("Errors should be 0", 0, error);
+      // Now wait for all transfers done
+      if (future2.isSuccess()) {
+        final DbTaskRunner runner2 = future2.getResult().getRunner();
+        logger.warn("Wait for second submit transfer {}", runner2);
+        if (runner2 != null) {
+          waitForAllDone(runner2);
+          dbTaskRunners.add(runner2);
+        }
+      }
+      Thread.sleep(100);
+    } else {
+      logger.error(Messages.getString("AbstractBusinessRequest.NeedMoreArgs",
+                                      "(-to -rule -file | -to -id)")); //$NON-NLS-1$
+      error++;
+    }
+    SubmitTransferTest.clearInternal();
+    String[] args2 = { "falsConfigFile", "-id", "1", "-to", "hostas" };
+    assertFalse(SubmitTransferTest.checkInternalParams(args2));
+    final long time2 = System.currentTimeMillis();
+    logger.warn("Success: " + success + " Error: " + error + " NB/s: " +
+                2 * success * 1000 / (time2 - time1));
+    Thread.sleep(1000);
+    totest.delete();
+    Thread.sleep(1000);
+    assertEquals("Success should be 2", 2, success);
     assertEquals("Errors should be 0", 0, error);
   }
 

--- a/doc/waarp-r66/source/changes.rst
+++ b/doc/waarp-r66/source/changes.rst
@@ -7,6 +7,17 @@ La procédure de mise à jour est disponible ici: :any:`upgrade`
 Non publié
 ==========
 
+Waarp R66 3.3.3 (2020-06-01)
+============================
+
+Correctifs
+==========
+
+- [`#31 <https://github.com/waarp/Waarp-All/pull/31>`__]
+  Corrige la régression sur la sélection d'un transfert à partir de son ID
+  où le nom du serveur local ne prenait pas en compte si le serveur
+  distant était en mode SSL ou pas (régression en 3.0).
+
 Waarp R66 3.3.3 (2020-05-07)
 ============================
 

--- a/pom.xml
+++ b/pom.xml
@@ -308,7 +308,7 @@
     <maven.resources.version>3.1.0</maven.resources.version>
     <maven.source.version>3.2.1</maven.source.version>
     <maven.dependency.version>3.1.2</maven.dependency.version>
-    <maven.site.version>3.9.0</maven.site.version>
+    <maven.site.version>3.8.2</maven.site.version>
     <maven.assembly.version>3.3.0</maven.assembly.version>
     <maven.javadoc.version>3.2.0</maven.javadoc.version>
     <maven.exec.version>1.6.0</maven.exec.version>
@@ -1383,6 +1383,7 @@
         <configuration>
           <skipEmptyReport>true</skipEmptyReport>
           <dependencyDetailsEnabled>true</dependencyDetailsEnabled>
+          <dependencyLocationsEnabled>false</dependencyLocationsEnabled>
           <!-- <skip>${module.project}</skip> -->
         </configuration>
         <reportSets>

